Migrates Flutter to null safety

vegaro opened the pull request:
I followed the steps in https://dart.dev/null-safety/migration-guide#step3-analyze and this PR is the result of the automatic migration.

We can publish this as version 3.0.0 of the SDK and it will show as a preview until dart 2.12.0 is out of beta.

Depends on https://github.com/RevenueCat/purchases-flutter/pull/154

aboedo commented:
it looks like the lint error will go away once this gets rebased from #154, so we should be good

aboedo commented on lib/entitlement_info_wrapper.dart:
```diff
@@ -38,13 +38,13 @@ class EntitlementInfo {
-  final String unsubscribeDetectedAt;
+  final String? unsubscribeDetectedAt;
 
   /// The date a billing issue was detected. Can be `nil` if there is no
```
we should probably update the end of this line to "can be `null` if there is no"

aboedo commented on lib/entitlement_info_wrapper.dart:
```diff
@@ -61,8 +61,8 @@ class EntitlementInfo {
-    var periodType;
-    switch (json["periodType"] as String) {
+    late var periodType;
+    switch (json["periodType"] as String?) {
```
what does this switch do if periodType is null? do we need to add a case to handle it?

IchordeDionysos replied:
If `periodType` is `null` then no case will match and periodType will still NOT be initialized.

And when it‘s then passed to `EntitlementInfo()` constructor as it‘s expected to be non-null it will crash.

There are three options:
- Leave it as it is and let it crash.
- Add a `case null: ...` to handle the null case.
- Make `periodType` nullable `String?` and be explicit that it may be null and handle it apprioriately.

I personally would prefer the last. So doing this:

```dart
String? periodType;
```

And the later:
```
EntitlementInfo(
  ...,
  periodType!,
  ...
  // Maybe this for the store (?!)
  store ?? Store.unknownStore,
);
```

I think you are better suited to judge what consequences would be acceptable.
Completely, preventing any errors while parsing is probably never possible.

aboedo commented on lib/entitlement_info_wrapper.dart:
```diff
@@ -73,8 +73,8 @@ class EntitlementInfo {
-    var store;
-    switch (json["store"] as String) {
+    late var store;
+    switch (json["store"] as String?) {
```
same question regarding switching on null

aboedo commented on lib/purchases_flutter.dart:
```diff
@@ -175,13 +175,13 @@ class Purchases {
       'packageIdentifier': packageToPurchase.identifier,
       'offeringIdentifier': packageToPurchase.offeringIdentifier,
       'oldSKU': upgradeInfo != null ? upgradeInfo.oldSKU : null,
       'prorationMode': upgradeInfo != null && upgradeInfo.prorationMode != null
```
do we need to be doing `upgradeInfo!.prorationMode` here?

IchordeDionysos replied:
It's not needed here as `upgradeInfo` is a function parameter and thus the analyzer can statically make sure that it's never `null`.

Unwrapping would cause a compilation error.

IchordeDionysos replied:
More information about this one:
https://github.com/dart-lang/sdk/issues/42626

aboedo replied:
that makes sense 👍

aboedo commented on lib/purchases_flutter.dart:
```diff
@@ -175,13 +175,13 @@ class Purchases {
       'oldSKU': upgradeInfo != null ? upgradeInfo.oldSKU : null,
       'prorationMode': upgradeInfo != null && upgradeInfo.prorationMode != null
-          ? upgradeInfo.prorationMode.index
+          ? upgradeInfo.prorationMode!.index
```
same question here, does this need to also be unwrapping `upgradeInfo`?

IchordeDionysos replied:
Here we do need unwrapping as `prorationMode` could be implemented like this:

```dart
String? get prorationMode {
  if (random() < 0.5) return null;
  return 'RevenueCat';
}
```

And the analyzer can't be sure that every time `upgradeInfo.prorationMode` is called the value is always null.

What I would suggest to do instead is this:

```dart
final prorationMode = upgradeInfo?.prorationMode;

prorationMode != null ? prorationMode.index : null
```

With that we don't need to assert that it's not null and the analyzer knows it itself.

aboedo replied:
agreed. @vegaro we should do the updates here, even if we're safe for now since we know the current implementation of upgradeInfo won't change values for `prorationMode` in between calls

aboedo reviewed (commented):
looking great! left a couple of questiosn

vegaro commented:
Added a couple tests for the store and purchaseType null when creating entitlement info, and fixed some casts that were changed by the migration tool, but that we actually don't need.

aboedo reviewed (approved):
@vegaro looks great! good to go once we improve safety on the changed calls to `proprationMode`. Thanks so much @IchordeDionysos for the detailed responses to my questions!!!

vegaro commented:
Sorry I missed that one @aboedo, I actually simplified oldSKU too.

vegaro commented:
Changed the base. This is meant to be a pre release until the final Dart version is released.

IchordeDionysos commented:
@vegaro the new Dart version with null-safety was just released 😍 
https://medium.com/dartlang/announcing-dart-2-12-499a6e689c87

So we can probably now release this 💪

vegaro commented:
I saw that! I am actually working on the new release now :)

vegaro commented:
@IchordeDionysos Just released 3.0.0

IchordeDionysos commented:
Very cool 😍  Thanks for keeping the RevenueCat Flutter SDK always so up-to-date 💪

vegaro commented:
Let us know if you face any issue or have any suggestion. And thanks so much for reviewing this. We really appreciate it
